File: 2021/12/22/main.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if head == None:
            return

        cur = head
        size = 0
        while cur != None:
            cur = cur.next
            size+=1

        cur = head
        for i in range((size-1)//2):
            cur = cur.next

        next = cur.next
        while next != None:
            temp = next.next
            next.next = cur
            cur = next
            next = temp
        end = cur

        while head != end:
            hnext = head.next
            enext = end.next
            head.next = end
            end.next = hnext
            head = hnext
            end = enext

        end.next = None

File: 2021/12/22/main.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reorderList(head *ListNode) {
	if head == nil {
		return
	}

	cur := head
	size := 0
	for cur != nil {
		cur = cur.Next
		size++
	}

	cur = head
	for i := 0; i < (size-1)/2; i++ {
		cur = cur.Next
	}

	next := cur.Next
	for next != nil {
		temp := next.Next
		next.Next = cur
		cur = next
		next = temp
	}
	end := cur

	for head != end {
		hnext := head.Next
		enext := end.Next
		head.Next = end
		end.Next = hnext
		head = hnext
		end = enext
	}

	end.Next = nil
}
